### PR TITLE
Isolate dtypes

### PR DESCRIPTION
**Pull request by CompRhys**
This is a fairly hefty set of changes that remove the use of global dtype setting from MACE codebase. This is done by adding custom `to` methods to change the dtypes of buffers. `dtype` is also made into a kwarg for all dataset objects.

All tests pass apart from those that have expected values that I believe are seed dependent and something must have been affected here that changes those values.

A bunch of minor code consistency changes are also included in this PR. Notable things are using `tmp_path` fixtures rather than manually creating and cleaning tmp folders, using the cli commands for the cli tests, adding some directories to the tests to make it more clear which tests are cli tests and which are core, adding concurrency to tests to reduce ci minute waste on PRs pushed in quick succession, using consistent docstring styles, using `npt.assert_allclose` consistently

The one major remaining source of confusion is in the `test_calculators.py` where using float32 causes everything to return very different/wrong values.

Related: https://github.com/ACEsuit/mace/issues/877, https://github.com/ACEsuit/mace/issues/328

### CODE REVIEW

File: mace/modules/radial.py
Comment (RokasEl) [inline mace/modules/radial.py:1]: Same here, with the `to` overrides. Here's an example of current behaviour:

```
from mace.modules.blocks import AtomicEnergiesBlock
import numpy as np
import torch

block = AtomicEnergiesBlock(atomic_energies=np.array([1.0, 2.0, 3.0]))
print(block.atomic_energies.dtype)
block = block.to(torch.float64)
print(block.atomic_energies.dtype)

>>>torch.float32
>>>torch.float64
```

File: mace/modules/utils.py [hunk 1]
```diff
@@ -302,8 +302,12 @@ def extract_invariant(x: torch.Tensor, num_layers: int, num_features: int, l_max
 def compute_mean_std_atomic_inter_energy(
     data_loader: torch.utils.data.DataLoader,
     atomic_energies: np.ndarray,
+    dtype: Optional[torch.dtype] = None,
 ) -> Tuple[float, float]:
-    atomic_energies_fn = AtomicEnergiesBlock(atomic_energies=atomic_energies)
+    dtype = dtype or torch.get_default_dtype()
+    atomic_energies_fn = AtomicEnergiesBlock(atomic_energies=atomic_energies).to(
+        dtype=dtype
+    )
```
Comment (RokasEl) [inline mace/modules/utils.py:310]: I like this way of setting the dtypes, but I don't think we need to override the `to` methods everywhere.

File: mace/modules/wrapper_ops.py [hunk 2]
```diff
@@ -150,8 +150,6 @@ def __new__(
                 layout=cueq_config.layout,
                 shared_weights=shared_weights,
                 internal_weights=internal_weights,
-                dtype=torch.get_default_dtype(),
-                math_dtype=torch.get_default_dtype(),
             )
         if (
             OEQ_AVAILABLE
```
Comment (RokasEl) [inline mace/modules/wrapper_ops.py:155]: I'm pretty sure the dtype is necessary here, as the kernels might be different depending on dtype

File: mace/tools/model_script_utils.py [hunk 3]
```diff
@@ -67,7 +71,7 @@ def configure_model(
 
     elif (args.mean is None or args.std is None) and args.model != "AtomicDipolesMACE":
         args.mean, args.std = modules.scaling_classes[args.scaling](
-            train_loader, atomic_energies
+            train_loader, atomic_energies, dtype=dtype
         )
```
Comment (RokasEl) [inline mace/tools/model_script_utils.py:75]: Do these classes really need a dtype kwarg? Can't this be done with a `cls.to(dtype)`?

File: mace/tools/torch_tools.py [hunk 4]
```diff
@@ -72,7 +77,12 @@ def init_device(device_str: str) -> torch.device:
     return torch.device("cpu")
 
 
-dtype_dict = {"float32": torch.float32, "float64": torch.float64}
+dtype_dict = {
+    "float32": torch.float32,
+    "float64": torch.float64,
+    "": torch.get_default_dtype(),
+    None: torch.get_default_dtype(),
+}
```
Comment (RokasEl) [inline mace/tools/torch_tools.py:85]: I like this!

File: mace/data/lmdb_dataset.py [hunk 5]
```diff
@@ -1,6 +1,10 @@
+from __future__ import annotations
```
[the remark this answers is not in the document]
Comment (CompRhys) [reply mace/data/lmdb_dataset.py:1]: MACE chooses to lint for py3.8 and non compatible type hints introduced later were used in this file

File: mace/cli/run_train.py [hunk 6]
```diff
@@ -157,29 +163,36 @@ def run(args) -> None:
             model_foundation = torch.load(
                 args.foundation_model, map_location=args.device
             )
+            model_foundation = model_foundation.to(dtype=dtype)
```
[the remark this answers is not in the document]
Comment (CompRhys) [reply mace/cli/run_train.py:166]: this is again likely path dependent from trying to get it to work in that I was littering these in lots of places, will try find the right place for a single call.

File: mace/modules/wrapper_ops.py (hunk 2 above)
Comment (CompRhys) [reply mace/modules/wrapper_ops.py:194]: this does turn out to be problematic which is maybe poor design from cuequivariance (especially given https://github.com/PASSIONLab/OpenEquivariance/issues/118 was responsive and implemented a working .to method that does update the kernels according to maintainers)

**Comment by CompRhys** [comment]
This is blocked by getting a solution to https://github.com/NVIDIA/cuEquivariance/issues/124 and then updating the values in various tests that assert against seeded answers.

**Comment by CompRhys** [comment]
`.to` method has been added in 0.6.0 cueq release so will try revive this at the weekend. It will require updating the pin to be `>0.6.0` for cueq which may cascade other issues.
